Update detached plugins to reflect security releases

jglick opened the pull request:
# Description

See below.

### Changelog entries

Proposed changelog entries:

* Updated the Mailer plugin version used when updating from Jenkins 1.x to include the fix of [SECURITY-372](https://jenkins.io/security/advisory/2017-03-20/), and the SSH Slaves plugin for [SECURITY-161](https://jenkins.io/security/advisory/2017-03-20/), and the Script Security plugin for [SECURITY-258](https://jenkins.io/security/advisory/2016-04-11/).

### Desired reviewers

@reviewbybees & @jenkinsci/code-reviewers

jglick committed:
[SECURITY-372] Update bundled mailer to 1.20.

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

daniel-beck commented:
AFAIUI (and this has come up with @oleg-nenashev in #2803 recently) upgrading the bundled version (which now only happens for important reasons) should generally also include a change to the `ClassicPluginStrategy` list, so that plugins get upgraded if they're outdated.

daniel-beck commented:
@jglick  Same test failures on all three platforms, so probably real.

jglick commented:
> upgrading the bundled version \[…\] should generally also include a change to the `ClassicPluginStrategy` list

This does not make sense. The `DETACHED` list is just supposed to show which versions were actually split off from core, and thus what the implied dependency should be (`getImpliedDependencies`). The code in the second clause of `PluginManager.loadDetachedPlugins`, apparently written by @stephenc in #2489, seems wrong. It should be forcing an upgrade when the installed version is older than that in `WEB-INF/detached-plugins/`.

In other words, there are two relevant versions for a detached plugin: the initial version after the split, reflecting the minimum API version for use in old compiled plugins which might have been using its classes from core; and the version we currently want users to install.

#2606 did not update `DETACHED_LIST` BTW.

jglick commented:
So to the test failure: this is about a new dependency on `display-url-api`. Does not look like `loadDetachedPlugins` as written will handle this, unless we added that plugin to `DETACHED_LIST`, which would not make any sense since it should not be an implied dependency of anything. Again the fix would be to ignore `getDetachedPlugins()` here and just examine `WEB-INF/detached-plugins/`.

jglick commented:
I suppose this could also bump up `ssh-slaves`.

jglick committed:
Merge branch 'master' into mailer-1.20
Picking up #2829 to see if that makes a difference in tests.

jglick committed:
Updated to pick up other detached plugins mentioned in https://jenkins.io/security/advisory/2016-04-11/ or https://jenkins.io/security/advisory/2017-03-20/ while we are here.

jglick commented:
`InstallerTranslatorTest` failures are being tracked in an INFRA issue. Others look real, will try to fix.

jglick committed:
I18nTest.test_baseName_plugin caught that the detached version of credentials was too old to serve as a dependency.
Probably there should be some test case verifying that all detached plugins could be loaded at once,
though it would need to do a topological sort in dependency order.

jglick committed:
Must also include display-url-api as a detached plugin as it is a dependency, even though it was never actually detached from core.

jglick commented:
@daniel-beck any comment on current state?

daniel-beck commented:
@jglick Looks good, assuming these versions work together.

---

>  Must also include display-url-api as a detached plugin as it is a dependency, even though it was never actually detached from core.

Similar to `credentials`, `ssh-credentials`, possibly others.

---

> apparently written by @stephenc in #2489, seems wrong. It should be forcing an upgrade when the installed version is older than

Did you find out what is going on here?

jglick commented:
> Did you find out what is going on here?

Not really; was hoping @stephenc would either agree that my TODO comment is on target and that there is something that should be fixed at some point, or explain what I am missing.

oleg-nenashev commented:
@jglick @daniel-beck @stephenc 

> Did you find out what is going on here?

IIRC there was an issue with upgrade over Jenkins 2.0 when I was looking into the code last time. For instance, if you upgrade from 1.642 to whatever release with #2489, upgrade of detached plugins does not get enforced due to going through the `InstallState.UPGRADE` branch in `PluginManager#loadDetachedPlugins`.

jglick commented:
IIUC this means that some sort of manual testing is required. Frankly I am not sure I follow what the _expected_ behavior during upgrades is, and what scenarios we are trying to fix. Is this about upgrade from ancient releases prior to the split? Or just routine upgrade of, say, 1.651.3 where these plugins were installed but in older versions and we want to upgrade them?

daniel-beck commented:
> s this about upgrade from ancient releases prior to the split? Or just routine upgrade of, say, 1.651.3 where these plugins were installed but in older versions and we want to upgrade them?

I *thought* the latter is what @stephenc 's PR did, but it doesn't (it's some very unusual edge case handling only AFAICT). Current behavior is the former, if a user upgrades past the detachment release, they get the bundled version, but if they are already past that, they stay on a possibly older reelase.

TBH, NBD IMO given the security warnings since earlier this year -- we now tell users about unsafe bundled plugins they have installed, and the bundled plugin installation via core upgrade handles it correctly (once this PR is merged).

So :+1: from me. Let's make at least the upgrades from ancient releases safe.

jglick commented:
@reviewbybees done

in that case.

oleg-nenashev commented on test/src/test/groovy/hudson/model/AbstractProjectTest.groovy:
```diff
@@ -555,6 +555,7 @@ public class AbstractProjectTest {
     public void testConfigDotXmlSubmissionToDifferentType() {
         j.jenkins.pluginManager.installDetachedPlugin("javadoc")
         j.jenkins.pluginManager.installDetachedPlugin("junit")
+        j.jenkins.pluginManager.installDetachedPlugin("display-url-api")
```
it is not a detached plugin, but yes we bundle it as a dependency :(

oleg-nenashev commented on war/pom.xml:
```diff
@@ -353,8 +353,14 @@ THE SOFTWARE.
                 <artifactItem>
                   <groupId>org.jenkins-ci.plugins</groupId>
+                  <artifactId>display-url-api</artifactId>
+                  <version>1.0</version>
```
🐛 It must be Display URL API 2.0 with removed JUnit dependency

jglick replied:
Why? Users do not care; the changes in 2.0 were done for architectural reasons. The plugin versions here are the minimum required to pick up all security advisories, plus the minimum transitive dependencies. Which in this case is 1.0. Certainly newer versions of plugins have all sorts of improvements and bug fixes that users may care about, and they are free to update to them at their leisure.

oleg-nenashev reviewed (changes requested):
🐛 for Display URL API 1.0. In this context we must not rely on a plugin pilling in extra dependencies

oleg-nenashev commented:
Apart from Display URL API, I agree that for security releases it makes sense to update the dependency. For other stuff I should create a follow-up JIRA ticket

oleg-nenashev commented:
@jglick Display URL API 2.0 is formally not binary compatible with 1.x. I do not want to advertise the obsolete version as a dependency.

jglick commented:
Formally yes but there is no practical impact.

daniel-beck commented:
@oleg-nenashev 

> I do not want to advertise the obsolete version as a dependency.

It's unclear what you mean here. This is just to allow upgrading users to get detached plugins with a dependency on display-url-api, it won't be a direct implied dependency of plugins depending on an older core release (no DETACHED_LIST entry).

oleg-nenashev commented:
> This is just to allow upgrading users to get detached plugins with a dependency on display-url-api, it won't be a direct implied dependency of plugins depending on an older core release (no DETACHED_LIST entry).

Yes, but we still declare the incompatible version in deps, which may impact stuff depending on this version, e.g. test suites. And it will be still pulling in the JUnit plugin in such case. I suggest keeping the latest version in this case unless there is a use-case to have Display URL API 1.x due to whatever reason in Jenkins.

jglick committed:
Merge branch 'master' into mailer-1.20

jglick committed:
@oleg-nenashev insists on bundling display-url-api 2.0.

oleg-nenashev reviewed (approved):
:bee: and @reviewbybees done

oleg-nenashev commented:
Going forward with merge, because it is a sensitive topic, which may require backporting. Better to have it tested in the wild.
